pkgs/development/compilers/llvm/common/bolt/default.nix

```diff
@@ -0,0 +1,71 @@
+{
+  lib,
+  stdenv,
+  llvm_meta,
+  monorepoSrc,
+  runCommand,
+  cmake,
+  libxml2,
+  libllvm,
+  libclang,
+  version,
+  python3,
+  buildLlvmTools,
+  patches ? [ ],
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "bolt";
+  inherit version patches;
+
+  # Blank llvm dir just so relative path works
+  src = runCommand "bolt-src-${finalAttrs.version}" { } ''
+    mkdir $out
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${finalAttrs.pname} "$out"
+    cp -r ${monorepoSrc}/third-party "$out"
+
+    # tablegen stuff, probably not the best way but it works...
+    cp -r ${monorepoSrc}/llvm/ "$out"
+    chmod -R +w $out/llvm
+  '';
+
+  sourceRoot = "${finalAttrs.src.name}/bolt";
+
+  nativeBuildInputs = [
+    cmake
+    python3
+  ];
+
+  buildInputs = [
+    libllvm
+    libclang
+    libxml2
+  ];
+
+  cmakeFlags = lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
+    (lib.cmakeFeature "LLVM_TABLEGEN_EXE" "${buildLlvmTools.llvm}/bin/llvm-tblgen")
+  ];
+
+  postUnpack = ''
+    chmod -R u+w -- $sourceRoot/..
+  '';
+
+  prePatch = ''
+    cd ..
+  '';
+
+  postPatch = ''
+    cd bolt
+  '';
+
+  outputs = [
+    "out"
+    "lib"
```

Comment:
I see only `.a` archive files in the lib directory. I think these should live in the `dev` output rather than the `lib` output, otherwise they end up being depended on by the `bin` (in this case `out`) output via $out/nix-support/propagated-build-inputs even though they're not needed.

There would be some care needed however to keep the libbolt_rt*.a in the same tree as bolt so it can find them.

This is handled in multiple-outputs.sh:

https://github.com/NixOS/nixpkgs/blob/1db1e3c46750e53fa3649bcd2893a248fe6de292/pkgs/build-support/setup-hooks/multiple-outputs.sh#L183

Here's what I currently see, and propagated-build-inputs contains `/nix/store/7d04yb047wzmyr832slq50l2k7y7rn46-bolt-19.0.0-unstable-2024-07-08-lib`.

```
/nix/store/7d04yb047wzmyr832slq50l2k7y7rn46-bolt-19.0.0-unstable-2024-07-08-lib
└── lib
    ├── libLLVMBOLTCore.a
    ├── libLLVMBOLTPasses.a
    ├── libLLVMBOLTProfile.a
    ├── libLLVMBOLTRewrite.a
    ├── libLLVMBOLTRuntimeLibs.a
    ├── libLLVMBOLTTargetAArch64.a
    ├── libLLVMBOLTTargetRISCV.a
    ├── libLLVMBOLTTargetX86.a
    └── libLLVMBOLTUtils.a
/nix/store/4l8sg5jq9vjnf112n910k46m5dr19i62-bolt-19.0.0-unstable-2024-07-08
├── bin
│   ├── llvm-bolt
│   ├── llvm-boltdiff
│   ├── llvm-bolt-heatmap
│   ├── merge-fdata
│   └── perf2bolt
├── lib
│   ├── libbolt_rt_hugify.a
│   └── libbolt_rt_instr.a
└── nix-support
    └── propagated-build-inputs
```

Reply:
> I see only `.a` archive files in the lib directory. I think these should live in the `dev` output rather than the `lib` output, otherwise they end up being depended on by the `bin` (in this case `out`) output via $out/nix-support/propagated-build-inputs even though they're not needed.

Yeah, I agree. Only got this PR to a buildable state, now we can make these kinds of fine adjustments :smile_cat:.

Reply:
Had to keep `libbolt_rt_hugify.a` and `libbolt_rt_instr.a` since they're needed at runtime.